### ISSUE TITLE
Gemma-3-12b-pt high grad_norm during continual pre-training

### ISSUE THREAD

**Issue by marseller** [huggingface/transformers#40460]
### System Info

- `transformers` version: 4.53.3
- Platform: Linux-6.6.87.2-microsoft-standard-WSL2-x86_64-with-glibc2.35
- Python version: 3.11.13
- Huggingface_hub version: 0.34.4
- Safetensors version: 0.6.2
- Accelerate version: 1.10.0
- Accelerate config:    not found
- DeepSpeed version: 0.17.4
- PyTorch version (accelerator?): 2.7.1+cu126 (NA)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: 
  "zero_optimization": {
      "stage": 3,
      "offload_optimizer": {
          "device": "cpu",
          "pin_memory": true
      },

### Who can help?

@zach-huggingface @SunMarc

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

I am performing domain-specific continual pre-training on Gemma-3-12B-PT. During the first 100 million tokens, I start encountering issues: the gradient_norm spikes to extremely high values, and the model’s outputs degrade, even when using a very small learning rate of 5e-7 and max_grad_norm 1.0. 
I am using exactly the same data and setup with LLaMA-3.1-8B, and in that case, the gradient_norm behaves as expected. The training is being conducted using the standard Trainer class, the only change was in learning rate, which does not directly affect the computed gradients. Both of the pre-trained models are using tokenizer bos and eos tokens for tokenization.

I am using accelerate to launch the training script and initialize deepspeed distributed training.
accelerate launch --config_file "deepspeed_config.yaml" train.py 

deepspeed_config.yaml:
compute_environment: LOCAL_MACHINE
debug: true
deepspeed_config:
  deepspeed_config_file: deepspeed_config.json
  zero3_init_flag: false
distributed_type: DEEPSPEED
downcast_bf16: 'no'
machine_rank: 0
main_training_function: main
num_machines: 1
num_processes: 4
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false

Gemma-3-12B grad_norm

<img width="1314" height="351" alt="Image" src="https://github.com/user-attachments/assets/af7b1e9d-6276-43fd-a3bd-19ae8bff480d" />

Llama-3.1-8B grad_norm

<img width="1300" height="336" alt="Image" src="https://github.com/user-attachments/assets/b40e43af-c29a-4296-a678-d7b7b072e96b" />

    if training_args.resume_from_checkpoint is not None:
        checkpoint_path = snapshot_download(
        repo_id=training_args.hub_model_id,
        allow_patterns=[f"{training_args.resume_from_checkpoint}/*"], 
        local_dir="/opt/ml/model/checkpoints",  
        )
        checkpoint = f"{checkpoint_path}/{training_args.resume_from_checkpoint}"
        model_args.model_name_or_path =  f"{checkpoint_path}/{training_args.resume_from_checkpoint}"

    model, tokenizer = create_and_prepare_model(model_args)
    model.config.use_cache = not training_args.gradient_checkpointing
    if training_args.gradient_checkpointing:
        training_args.gradient_checkpointing_kwargs = {"use_reentrant": model_args.use_reentrant}

    train_dataset, eval_dataset = create_datasets_with_tokenizer(tokenizer, data_args)
    data_collator = DataCollatorWithFlattening()
    trainer = Trainer(
        model=model,
        data_collator=data_collator,
        tokenizer=tokenizer,
        args=training_args,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset,
    )
    trainer.accelerator.print(f"{trainer.model}")
    if hasattr(trainer.model, "print_trainable_parameters"):
        trainer.model.print_trainable_parameters()

    print(f"Starting training with model: {model_args.model_name_or_path} {checkpoint}")
    trainer.train(resume_from_checkpoint=checkpoint)

    if trainer.is_fsdp_enabled:
        trainer.accelerator.state.fsdp_plugin.set_state_dict_type("FULL_STATE_DICT")
    trainer.save_model()`

### Expected behavior

lower gradient updates for Gemma-3-12b-pt

**Comment by marseller** [comment]
related to  #40564

**Comment by marseller** [comment]
I have reopened the issue because the loss and grad_norm currently are correctly scaled , but gradient spikes still occur in gemma-3-12b-pt continual pre-training.

<img width="1312" height="389" alt="Image" src="https://github.com/user-attachments/assets/f3f76e3f-35d4-4fcc-a5ec-763925633318" />

I have max_grad_norm 1.0 enabled during training, so the spikes don't affect the loss, however inferencing the model and testing it on benchmark shows degraded performance.

**Comment by Rocketknight1** [comment]
Hmmn, this is unlikely to be a bug in `transformers` since I think our implementation of Gemma3 is correct, or else we'd see errors in inference. More likely, this architecture is prone to gradient spikes in some places; this is common in a lot of architectures, especially larger ones. You may have to refer to the paper or experiment with techniques like gradient clipping etc. I don't think we can diagnose this issue for you since it only appears in a longer training run!
